File: coco/pattern/sensec/app/services/encryption_service.py
from __future__ import annotations

import base64
import uuid

from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.audit.logger import audit_log
from app.auth.pbac_client import NullPBACClient, PBACClient
from app.crypto import dek_manager
from app.crypto.dek_manager import make_fingerprint, pack_token
from app.crypto.kek_client import KEKClient
from app.models.edek_record import EDEKRecord, RotationStatus
from app.models.schemas import EncryptRequest, EncryptResponse


async def encrypt(
    request: EncryptRequest,
    app_id: str,
    caller_sub: str,
    kek_client: KEKClient,
    session: AsyncSession,
    caller_ip: str = "",
    pbac_client: PBACClient | NullPBACClient | None = None,
) -> EncryptResponse:
    if pbac_client is not None and request.end_user_id:
        permitted = await pbac_client.check(
            end_user_id=request.end_user_id,
            action="encrypt",
            data_classification=request.data_classification,
            context={"app_id": app_id, "caller_ip": caller_ip},
        )
        if not permitted:
            audit_log(
                "encrypt",
                app_id=app_id,
                sub=caller_sub,
                end_user_id=request.end_user_id,
                caller_ip=caller_ip,
                status="failure",
                reason="pbac_denied",
            )
            raise HTTPException(status.HTTP_403_FORBIDDEN, "Access denied by policy")

    dek = dek_manager.generate_dek()
    try:
        result = dek_manager.encrypt(request.plaintext.encode(), dek, app_id)
        edek_bytes, kek_version = await kek_client.wrap_dek(bytes(dek))
    finally:
        dek_manager.zero_dek(dek)

    record = EDEKRecord(
        edek_id=uuid.uuid4(),
        app_id=app_id,
        edek_blob=base64.b64encode(edek_bytes).decode(),
        kek_version=kek_version,
        algorithm=dek_manager.ALGORITHM,
        encoding=request.encoding,
        data_classification=request.data_classification,
        rotation_status=RotationStatus.current,
        fingerprint=make_fingerprint(result.iv, result.tag),
    )
    session.add(record)
    await session.commit()

    audit_log(
        "encrypt",
        app_id=app_id,
        sub=caller_sub,
        end_user_id=request.end_user_id,
        edek_id=str(record.edek_id),
        kek_version=kek_version,
        data_classification=request.data_classification,
        caller_ip=caller_ip,
        context=request.context,
        status="success",
    )

    return EncryptResponse(
        ciphertext_token=pack_token(record.edek_id, result.iv, result.tag, result.ciphertext),
        edek_id=record.edek_id,
        owner_app_id=app_id,
        algorithm=dek_manager.ALGORITHM,
        encoding=request.encoding,
        kek_version=kek_version,
        # deprecated individual fields — kept for backward compatibility
        iv_b64=base64.b64encode(result.iv).decode(),
        ciphertext_b64=base64.b64encode(result.ciphertext).decode(),
        tag_b64=base64.b64encode(result.tag).decode(),
    )

File: coco/pattern/springsec/hsm-encryption-service/src/main/java/com/hsm/encryption/service/EncryptionService.java
package com.hsm.encryption.service;

import com.hsm.encryption.audit.AuditLogger;
import com.hsm.encryption.auth.PbacClient;
import com.hsm.encryption.crypto.DekManager;
import com.hsm.encryption.crypto.KekClient;
import com.hsm.encryption.dto.EncryptRequest;
import com.hsm.encryption.dto.EncryptResponse;
import com.hsm.encryption.model.EdekRecord;
import com.hsm.encryption.repository.EdekRecordRepository;
import com.hsm.encryption.web.ApiException;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Map;
import java.util.UUID;

/** Ported from app/services/encryption_service.py. */
@Service
public class EncryptionService {

    private static final int MAX_PLAINTEXT_BYTES = EncryptRequest.MAX_PLAINTEXT_CHARS;

    private final KekClient kekClient;
    private final EdekRecordRepository edekRecordRepository;
    private final PbacClient pbacClient;
    private final AuditLogger auditLogger;

    public EncryptionService(KekClient kekClient, EdekRecordRepository edekRecordRepository,
                              PbacClient pbacClient, AuditLogger auditLogger) {
        this.kekClient = kekClient;
        this.edekRecordRepository = edekRecordRepository;
        this.pbacClient = pbacClient;
        this.auditLogger = auditLogger;
    }

    public EncryptResponse encrypt(EncryptRequest request, String appId, String callerSub, String callerIp) {
        byte[] plaintextBytes = request.plaintext().getBytes(StandardCharsets.UTF_8);
        if (plaintextBytes.length > MAX_PLAINTEXT_BYTES) {
            throw new ApiException(HttpStatus.UNPROCESSABLE_ENTITY,
                    "plaintext exceeds maximum size: " + plaintextBytes.length
                            + " bytes (hard limit " + MAX_PLAINTEXT_BYTES + " bytes)");
        }

        if (request.endUserId() != null && !request.endUserId().isBlank()) {
            boolean permitted = pbacClient.check(
                    request.endUserId(), "encrypt", request.dataClassification(),
                    Map.of("app_id", appId, "caller_ip", nullToEmpty(callerIp))
            );
            if (!permitted) {
                auditLogger.log("encrypt",
                        "app_id", appId, "sub", callerSub, "end_user_id", request.endUserId(),
                        "caller_ip", callerIp, "status", "failure", "reason", "pbac_denied");
                throw new ApiException(HttpStatus.FORBIDDEN, "Access denied by policy");
            }
        }

        byte[] dek = DekManager.generateDek();
        DekManager.EncryptResult result;
        KekClient.WrapResult wrapResult;
        try {
            result = DekManager.encrypt(plaintextBytes, dek, appId);
            wrapResult = kekClient.wrapDek(dek);
        } finally {
            DekManager.zeroDek(dek);
        }

        UUID edekId = UUID.randomUUID();
        String fingerprint = DekManager.makeFingerprint(result.iv(), result.tag());
        EdekRecord record = new EdekRecord(
                edekId, appId, Base64.getEncoder().encodeToString(wrapResult.edekBytes()), wrapResult.kekVersion(),
                DekManager.ALGORITHM, request.encoding(), request.dataClassification(), fingerprint
        );
        edekRecordRepository.save(record);

        auditLogger.log("encrypt",
                "app_id", appId, "sub", callerSub, "end_user_id", request.endUserId(),
                "edek_id", edekId.toString(), "kek_version", wrapResult.kekVersion(),
                "data_classification", request.dataClassification(), "caller_ip", callerIp,
                "context", request.context(), "status", "success");

        return new EncryptResponse(
                DekManager.packToken(edekId, result.iv(), result.tag(), result.ciphertext()),
                edekId, appId, DekManager.ALGORITHM, request.encoding(), wrapResult.kekVersion(),
                Base64.getEncoder().encodeToString(result.iv()),
                Base64.getEncoder().encodeToString(result.ciphertext()),
                Base64.getEncoder().encodeToString(result.tag())
        );
    }

    private static String nullToEmpty(String s) {
        return s == null ? "" : s;
    }
}
